opal/mca/base/mca_base_component_repository.c

```diff
@@ -74,321 +59,448 @@ static bool initialized = false;
-       component that is depended on by other components actually
-       causes missing symbols because the OS actually does unload it
-       from memory!), such as OS X.
-
-       So instead of just blindly closing everything, we have iterate
-       over the array of open components releasing everything with a
-       refcount of 1 -- skip anything with a refcount of more than 1.
-       Repeat this procedure until either we have nothing open or we
-       made one full pass and no refcounts went to 1 (which is
-       technically an error). */
+    /* Initialize the dl framework */
+    int ret = mca_base_framework_open(&opal_dl_base_framework, 0);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(0, "%s %d:%s failed -- process will likely abort (open the dl framework returned %d instead of OPAL_SUCCESS)\n",
+                    __FILE__, __LINE__, __func__, ret);
+        return ret;
+    }
+    opal_dl_base_select();
 
-    do {
-      OPAL_LIST_FOREACH_SAFE(ri, next, &repository, repository_item_t) {
-        OBJ_RELEASE(ri);
-      }
-    } while (opal_list_get_size(&repository) > 0);
+    OBJ_CONSTRUCT(&mca_base_component_repository, opal_hash_table_t);
+    ret = opal_hash_table_init (&mca_base_component_repository, 128);
+    if (OPAL_SUCCESS != ret) {
+        mca_base_framework_close (&opal_dl_base_framework);
+        return ret;
+    }
 
-    (void) mca_base_framework_close(&opal_dl_base_framework);
+    ret = mca_base_component_repository_add (mca_base_component_path);
+    if (OPAL_SUCCESS != ret) {
+        OBJ_DESTRUCT(&mca_base_component_repository);
+        mca_base_framework_close (&opal_dl_base_framework);
+        return ret;
+    }
 #endif
 
-    initialized = false;
+    initialized = true;
   }
+
+  /* All done */
+
+  return OPAL_SUCCESS;
 }
 
+int mca_base_component_repository_get_components (mca_base_framework_t *framework,
+                                                  opal_list_t **framework_components)
+{
+    *framework_components = NULL;
 #if OPAL_HAVE_DL_SUPPORT
+    return opal_hash_table_get_value_ptr (&mca_base_component_repository, framework->framework_name,
+                                          strlen (framework->framework_name), (void **) framework_components);
+#endif
+    return OPAL_ERR_NOT_FOUND;
+}
 
-static repository_item_t *find_component(const char *type, const char *name)
+void mca_base_component_repository_release(const mca_base_component_t *component)
 {
-  opal_list_item_t *item;
-  repository_item_t *ri;
-
-  for (item = opal_list_get_first(&repository);
-       opal_list_get_end(&repository) != item;
-       item = opal_list_get_next(item)) {
-    ri = (repository_item_t *) item;
-    if (0 == strcmp(ri->ri_type, type) && 
-        0 == strcmp(ri->ri_component_struct->mca_component_name, name)) {
-      return ri;
-    }
-  }
-
-  /* Not found */
-
-  return NULL;
+    /* TODO -- should we really release the component. it might get selected next time */
```

Comment:
What will you do about this?  Should probably resolve all TODO's before merging this PR.